api4/bot.go

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package api4
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+func (api *API) InitBot() {
+	api.BaseRoutes.Bots.Handle("", api.ApiHandler(createBot)).Methods("POST")
+	api.BaseRoutes.Bot.Handle("", api.ApiSessionRequired(patchBot)).Methods("PUT")
+	api.BaseRoutes.Bot.Handle("", api.ApiSessionRequired(getBot)).Methods("GET")
+	api.BaseRoutes.Bots.Handle("", api.ApiSessionRequired(getBots)).Methods("GET")
+	api.BaseRoutes.Bot.Handle("/disable", api.ApiSessionRequired(disableBot)).Methods("POST")
+}
+
+func sessionHasPermissionToManageBot(c *Context, userId string) *model.AppError {
+	existingBot, err := c.App.GetBot(c.Params.UserId, true)
+	if err != nil {
+		return err
+	}
+
+	if existingBot.CreatorId == c.App.Session.UserId {
+		if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_BOTS) {
+			return c.MakePermissionError(model.PERMISSION_MANAGE_BOTS)
+		}
+	} else {
+		if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_OTHERS_BOTS) {
+			return c.MakePermissionError(model.PERMISSION_MANAGE_OTHERS_BOTS)
+		}
+	}
+
+	return nil
+}
+
+func createBot(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_CREATE_BOT) {
+		c.SetPermissionError(model.PERMISSION_CREATE_BOT)
+		return
+	}
+
+	botPatch := model.BotPatchFromJson(r.Body)
+	if botPatch == nil {
+		c.SetInvalidParam("bot")
+		return
+	}
+
+	bot := &model.Bot{
+		CreatorId: c.App.Session.UserId,
+	}
+	bot.Patch(botPatch)
+
+	createdBot, err := c.App.CreateBot(bot)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	w.Write(createdBot.ToJson())
+}
+
+func patchBot(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireUserId()
+	if c.Err != nil {
+		return
+	}
+
+	if err := sessionHasPermissionToManageBot(c, c.Params.UserId); err != nil {
+		c.Err = err
+		return
+	}
+
+	botPatch := model.BotPatchFromJson(r.Body)
+	if botPatch == nil {
+		c.SetInvalidParam("bot")
+		return
+	}
+
+	updatedBot, err := c.App.PatchBot(c.Params.UserId, botPatch)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Write(updatedBot.ToJson())
+}
+
+func getBot(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireUserId()
+	if c.Err != nil {
+		return
+	}
+
+	includeDeleted := r.URL.Query().Get("include_deleted") == "true"
+
+	bot, err := c.App.GetBot(c.Params.UserId, includeDeleted)
+	if err != nil {
+		c.Err = err
```

Comment:
Agreed on the conclusion, and I note that even if we held the error back, we could likely still glean information from timing measurements to determine if the SQL API ran or not.

Reply:
@grundleborg, can you shed more light into the expected hardening that will come with guest accounts? 

So long as username remains a unique column, it seems like I can always infer the availability of a username by trying to change mine or create a new one as such. So the additional surface area here is that I can test if that username corresponds to a bot or a regular user. 

I think the right thing to do here is to fail with a 404 instead of a permission error if we try to read a bot that does exist but that doesn't belong to us. To do this consistently, it probably makes sense to propagate the `creatorId` matching the current session's `userId` down to the store just like we do for fetching all bots: then the error code for a permission mismatch will be identical to when it really doesn't exist as a bot. 

This also means updating `patchBot` and `disableBot`. Instead of explicit permission checks in the `api4` layer, we'd now effectively have the store helping us by removing rows we shouldn't access.

However, if a user has permission to create their own bots, we're back to their being able to test if a username corresponds to a bot or not by trying to create a bot with that username. What's your advice on where to draw the lines for hardening here?

Reply:
My instinct here would be that worrying about timing attacks here is futile because (a) the user already has logged-in access to the server, and (b) because our architecture makes it hard to defend against in the case of (a).

I'm about 1/5 that we should either return the same error code in both cases, or alternatively, return the same error code *in the case where the experimental security hardening config flag* is enabled. Happy to discuss in more detail / loop in Security Meta Team / whatever though.

Reply:
Sounds good -- I'll raise this thread with the security meta team.